File: frontend/src/types/entities/stock-count.ts
export type StockCountMode =
  | "blind"
  | "visible";

export type StockCountSourceType =
  | "snapshot"
  | "discovered";


export interface StockCountWarehouse {
  id: string;

  code: string;

  name: string;
}


export interface StockCountUser {
  id: string;

  name: string | null;

  username: string | null;
}


export interface StockCountProduct {
  id: string;

  internal_sku: string;

  name: string;

  track_batches: boolean;

  track_expiry: boolean;
}


export type StockCountScopeResolutionStatus =
  | "unresolved"
  | "physical_lines"
  | "no_stock_confirmed";


export interface StockCountScopeProduct {
  product: StockCountProduct;

  resolution_status: StockCountScopeResolutionStatus;

  /*
   * Number of physical Stock Count identities currently represented
   * by snapshot or discovered lines for this selected Product.
   *
   * This is a line count, not a stock quantity.
   */
  physical_line_count: number;

  no_stock_confirmed_at: string | null;

  no_stock_confirmed_by: StockCountUser | null;
}


export interface StockCountBatch {
  id: string;

  batch_number: string | null;

  expiry_date: string | null;

  is_expired: boolean;
}


export interface StockCountSummary {
  total_items: number;

  counted_items: number;

  uncounted_items: number;

  /*
   * Variance fields are intentionally absent while an open
   * blind count is in progress.
   */
  variance_items?: number;

  positive_variance_items?: number;

  negative_variance_items?: number;
}


export interface StockCountAdjustmentLink {
  id: string;

  adjustment_number: string;
}


export interface StockCountItem {
  id: string;

  line_number: number;

  source_type: StockCountSourceType;

  product: StockCountProduct;

  /*
   * Canonical inventory batch attached to snapshot lines.
   *
   * Discovered lines initially have no InventoryBatch and
   * therefore return null here.
   */
  batch: StockCountBatch | null;

  observed_batch_number: string | null;

  observed_expiry_date: string | null;

  /*
   * System-derived quantities are deliberately omitted from
   * open blind-count responses.
   */
  snapshot_quantity?: string;

  expected_quantity?: string;

  counted_quantity: string | null;

  variance_quantity?: string | null;

  counted_at: string | null;

  counted_by: StockCountUser | null;

  notes: string | null;
}


export interface StockCount {
  id: string;

  count_number: string;

  status:
    | "open"
    | "completed"
    | "cancelled";

  scope_type:
    | "full"
    | "selected";

  count_mode: StockCountMode;

  warehouse: StockCountWarehouse;

  snapshot_at: string | null;

  started_at: string | null;

  started_by: StockCountUser | null;

  completed_at: string | null;

  completed_by: StockCountUser | null;

  cancelled_at: string | null;

  cancelled_by: StockCountUser | null;

  notes: string | null;

  adjustment: StockCountAdjustmentLink | null;

  summary: StockCountSummary;

  /*
   * Explicit selected Product membership.
   *
   * Full Warehouse counts return an empty array.
   */
  scope_products: StockCountScopeProduct[];

  items: StockCountItem[];

  created_at: string | null;

  updated_at: string | null;
}

File: app/serializers/stock_count.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from app.models import InventoryBatch, Product, StockAdjustment, StockCount, StockCountItem, Warehouse


def _timestamp(value) -> str | None:
    return value.isoformat() if value else None


def _date(value) -> str | None:
    return value.isoformat() if value else None


def _decimal(value) -> str | None:
    if value is None:
        return None
    if isinstance(value, Decimal):
        return str(value)
    return str(value)


def _user(value: dict | None) -> dict | None:
    if not value:
        return None
    return {
        "id": str(value["id"]),
        "name": " ".join(
            part
            for part in [
                value["first_name"],
                value["last_name"],
            ]
            if part
        )
        or None,
        "username": value["username"],
    }


def _is_expired(expiry_date: date | None) -> bool:
    return expiry_date is not None and expiry_date < date.today()


def stock_count_summary(
    *,
    total_items: int,
    counted_items: int,
    variance_items: int,
    positive_variance_items: int,
    negative_variance_items: int,
) -> dict:
    return {
        "total_items": total_items,
        "counted_items": counted_items,
        "uncounted_items": max(total_items - counted_items, 0),
        "variance_items": variance_items,
        "positive_variance_items": positive_variance_items,
        "negative_variance_items": negative_variance_items,
    }


def serialize_stock_count_item(
    item: StockCountItem,
    *,
    product: Product,
    batch: InventoryBatch | None,
    counted_by: dict | None,
) -> dict:
    return {
        "id": str(item.id),
        "line_number": item.line_number,
        "product": {
            "id": str(product.id),
            "internal_sku": product.internal_sku,
            "name": product.name,
            "track_batches": bool(product.track_batches),
            "track_expiry": bool(product.track_expiry),
        },
        "batch": (
            {
                "id": str(batch.id),
                "batch_number": batch.batch_number,
                "expiry_date": _date(batch.expiry_date),
                "is_expired": _is_expired(batch.expiry_date),
            }
            if batch
            else None
        ),
        "snapshot_quantity": _decimal(item.snapshot_quantity),
        "expected_quantity": _decimal(item.expected_quantity),
        "counted_quantity": _decimal(item.counted_quantity),
        "variance_quantity": _decimal(item.variance_quantity),
        "counted_at": _timestamp(item.counted_at),
        "counted_by": _user(counted_by),
        "notes": item.notes,
    }


def serialize_stock_count(
    count: StockCount,
    *,
    warehouse: Warehouse,
    started_by: dict | None,
    completed_by: dict | None,
    cancelled_by: dict | None,
    items: list[tuple[StockCountItem, Product, InventoryBatch | None, dict | None]],
    adjustment: StockAdjustment | None = None,
) -> dict:
    serialized_items = [
        serialize_stock_count_item(
            item,
            product=product,
            batch=batch,
            counted_by=counted_by,
        )
        for item, product, batch, counted_by in items
    ]
    counted_items = [
        item
        for item in serialized_items
        if item["counted_quantity"] is not None
    ]
    variance_items = [
        item
        for item in counted_items
        if Decimal(item["variance_quantity"] or "0") != Decimal("0")
    ]

    return {
        "id": str(count.id),
        "count_number": count.count_number,
        "status": count.status,
        "scope_type": count.scope_type,
        "warehouse": {
            "id": str(warehouse.id),
            "code": warehouse.code,
            "name": warehouse.name,
        },
        "snapshot_at": _timestamp(count.snapshot_at),
        "started_at": _timestamp(count.started_at),
        "started_by": _user(started_by),
        "completed_at": _timestamp(count.completed_at),
        "completed_by": _user(completed_by),
        "cancelled_at": _timestamp(count.cancelled_at),
        "cancelled_by": _user(cancelled_by),
        "notes": count.notes,
        "adjustment": (
            {
                "id": str(adjustment.id),
                "adjustment_number": adjustment.adjustment_number,
            }
            if adjustment
            else None
        ),
        "summary": stock_count_summary(
            total_items=len(serialized_items),
            counted_items=len(counted_items),
            variance_items=len(variance_items),
            positive_variance_items=len(
                [
                    item
                    for item in variance_items
                    if Decimal(item["variance_quantity"] or "0") > Decimal("0")
                ]
            ),
            negative_variance_items=len(
                [
                    item
                    for item in variance_items
                    if Decimal(item["variance_quantity"] or "0") < Decimal("0")
                ]
            ),
        ),
        "items": serialized_items,
        "created_at": _timestamp(count.created_at),
        "updated_at": _timestamp(count.updated_at),
    }


def serialize_stock_count_summary(
    count: StockCount,
    *,
    warehouse: Warehouse,
    started_by: dict | None,
    item_counts: dict,
    adjustment: StockAdjustment | None = None,
) -> dict:
    return {
        "id": str(count.id),
        "count_number": count.count_number,
        "status": count.status,
        "scope_type": count.scope_type,
        "warehouse": {
            "id": str(warehouse.id),
            "code": warehouse.code,
            "name": warehouse.name,
        },
        "snapshot_at": _timestamp(count.snapshot_at),
        "started_at": _timestamp(count.started_at),
        "started_by": _user(started_by),
        "completed_at": _timestamp(count.completed_at),
        "cancelled_at": _timestamp(count.cancelled_at),
        "notes": count.notes,
        "adjustment": (
            {
                "id": str(adjustment.id),
                "adjustment_number": adjustment.adjustment_number,
            }
            if adjustment
            else None
        ),
        "summary": stock_count_summary(**item_counts),
        "created_at": _timestamp(count.created_at),
        "updated_at": _timestamp(count.updated_at),
    }
